zerocopy-derive/src/enums.rs

```diff
@@ -0,0 +1,550 @@
+
+    let match_arms = data.variants.iter().map(|variant| {
+        let value_ident = idents.discriminant_const(&variant.ident);
+
+        if matches!(variant.fields, Fields::Unit) {
+            // Unit variants don't need any further validation beyond checking
+            // the discriminant.
+            quote! {
+                #value_ident => true
+            }
+        } else {
+            let variant_struct = idents.variant_struct(&variant.ident);
+
+            quote! {
+                #value_ident => {
+                    // SAFETY:
+                    // - `cast_unsized`:
+                    //   - The variant struct has the same size as the enum
+                    //   - The returned pointer is cast from `p`, and so has the
+                    //     same provenance as `p`.
+                    //   - We checked that the discriminant of the enum matched
+                    //     the value for this variant, so this cast preserves
+                    //     types and locations of all fields. Therefore, any
+                    //     `UnsafeCell`s will have the same location as in the
+                    //     original type.
+                    // - `assume_initialized`: `cast_unsized` removes the
+                    //   initialization invariant from `p`, so we re-assert that
+                    //   all of the bytes are initialized.
+                    let candidate = unsafe {
+                        fields
+                            .cast_unsized(|p: *mut #fields_ident| p as *mut #variant_struct)
+                            .assume_initialized()
+                    };
+                    <#variant_struct>::is_bit_valid(candidate)
+                }
+            }
+        }
+    });
+
+    quote! {
+        // SAFETY: We use `is_bit_valid` to validate that the bit pattern
+        // corresponds to one of the enum's variant discriminants. Then, we
+        // check the bit validity of each field of that variant. Thus, this is a
+        // sound implementation of `is_bit_valid`.
+        fn is_bit_valid<A>(
+            mut candidate: ::zerocopy::Maybe<'_, Self, A>,
+        ) -> ::zerocopy::macro_util::core_reexport::primitive::bool
+        where
+            A: ::zerocopy::pointer::invariant::Aliasing
+                + ::zerocopy::pointer::invariant::AtLeast<::zerocopy::pointer::invariant::Shared>,
+        {
+            use ::zerocopy::macro_util::core_reexport;
+
+            let candidate = candidate.reborrow().forget_exclusive();
+
+            #discriminant_type
+            #discriminant_consts
+            #(#variant_structs)*
+
+            #[repr(C)]
+            #[allow(non_snake_case)]
+            union #fields_ident #impl_generics #where_clause {
+                #(
+                    pub #variant_idents: core_reexport::mem::ManuallyDrop<
+                        #variant_struct_idents #ty_generics,
+                    >,
+                )*
+                __nonempty: (),
```

Comment:
I presume Rust doesn't permit field-less unions? Can you add a comment explaining that?

Reply:
Done